modepy/quadrature/finding.py

```diff
@@ -0,0 +1,422 @@
+
+    def __call__(self, points: np.ndarray) -> np.ndarray:
+        """Evaluate the linear combination of :attr:`functions` with
+        :attr:`coefficients` at *points*.
+        """
+        return sum(
+            (coeff * func(points)
+                for coeff, func in zip(self.coefficients, self.functions, strict=True)),
+            np.zeros(()))
+
+
+def linearly_combine(
+            coefficients: np.ndarray,
+            functions: Sequence[Integrand]
+        ) -> LinearCombinationIntegrand:
+    """
+    Takes advantage of associativity when linearly combining linear combinations.
+    """
+
+    lcfunctions = [
+        f for f in functions
+        if isinstance(f, LinearCombinationIntegrand)
+    ]
+    if len(lcfunctions) != len(functions):
+        return LinearCombinationIntegrand(coefficients, functions)
+
+    basis: list[Integrand] = []
+    n = len(lcfunctions)
+    matrix = np.zeros((n, n), dtype=np.complex128)
+    for i, f in enumerate(lcfunctions):
+        ncommon = min(len(basis), len(f.functions))
+        assert basis[:ncommon]  == f.functions[:ncommon]
+        basis.extend(f.functions[ncommon:])
+
+        ncoeff = len(f.coefficients)
+        matrix[i, :ncoeff] = f.coefficients
+
+    return LinearCombinationIntegrand(coefficients @ matrix, basis)
+
+
+def _mass_matrix(
+            integrate: Callable[[Integrand], np.inexact],
+            basis: Sequence[Integrand],
+        ) -> np.ndarray:
+    n = len(basis)
+    integrals = [
+        [
+            integrate(_ProductIntegrand((basis[i], _ConjugateIntegrand(basis[j]))))
+            for j in range(i+1)
+            ]
+        for i in range(n)
+    ]
+
+    # Let np.array deal with figuring out the result dtype
+    return np.array([
+        # Fill out the full matrix from the triangle computed above
+        [integrals[i][j] if i >= j else integrals[j][i] for j in range(n)]
+        for i in range(n)
+    ])
+
+
+def orthogonalize_basis(
+            integrate: Callable[[Integrand], np.inexact],
+            basis: Sequence[Integrand],
+        ) -> Sequence[LinearCombinationIntegrand]:
+    r"""
+    Let :math:`\Omega\subset\mathbb C^n` be a domain. (Domains
+    over the reals are allowable as well.) Returns linear combinations
+    of functions in *basis* that is orthogonal under the (complex-valued)
+    :math:`L^2` inner product induced by *integrate*.
+
+    :arg integrate: Computes an integral of the passed integrand over
+        :math:`\Omega`. Must use integration nodes compatible with
+        *basis*.
+    :arg basis: A sequence of functions that accept an array of nodes
+        of shape either ``(ndim, nnodes)`` or ``(nnodes,)`` and return
+        an array of shape ``(nnodes,)`` with the value of the
+        basis function at the node.
+    """
+    n = len(basis)
+    mass_mat = _mass_matrix(integrate, basis)
+
+    l_factor = la.cholesky(mass_mat)
+
+    from scipy.linalg import solve_triangular
```

Comment:
This should add some dependency on scipy in pyproject?

Reply:
Grr, you're right. And: no, I'm not doing that. I'd rather fall back to `la.solve`, even if it's asymptotically wrong.